src/main/java/org/broadinstitute/hellbender/utils/haplotype/AllHaplotypeBAMWriter.java

```diff
@@ -0,0 +1,45 @@
+package org.broadinstitute.hellbender.utils.haplotype;
+
+import org.broadinstitute.hellbender.utils.GenomeLoc;
+import org.broadinstitute.hellbender.utils.genotyper.ReadLikelihoods;
+import org.broadinstitute.hellbender.utils.haplotype.Haplotype;
+//import org.broadinstitute.hellbender.utils.sam.GATKSAMRecord;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * A haplotype bam writer that writes out all haplotypes as reads and then
+ * the alignment of reach read to its best match among the best haplotypes.
+ *
+ * Primarily useful for people working on the HaplotypeCaller method itself
+ */
+class AllHaplotypeBAMWriter extends HaplotypeBAMWriter {
+
+    public AllHaplotypeBAMWriter(final HaplotypeBAMDestination destination) {
+        super(destination);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void writeReadsAlignedToHaplotypes(final Collection<Haplotype> haplotypes,
+                                              final GenomeLoc paddedReferenceLoc,
+                                              final Collection<Haplotype> bestHaplotypes,
+                                              final Set<Haplotype> calledHaplotypes,
+                                              final ReadLikelihoods<Haplotype> readLikelihoods) {
+
+        writeHaplotypesAsReads(haplotypes, new HashSet<>(bestHaplotypes), paddedReferenceLoc);
+
+        final int sampleCount = readLikelihoods.numberOfSamples();
+        for (int i = 0; i < sampleCount; i++) {
+            for (final GATKRead read : readLikelihoods.sampleReads(i)) {
+                writeReadAgainstHaplotype(read);
+            }
+        }
+        output.close();
```

Comment:
this does not look like the right place to close this stream - it would blow up if someone wrote to it after this and it would be unexpected. I think a close() method on the superclass is what we need. Close needs to be called in a finally block or in try-with-resources